### PR TITLE
#1 Bug Fix: Sending Messages

### PR DESCRIPTION
**Pull request by xynree**
## Description

Sending messages now loads properly and messages appear in order they were sent.

## Notes 

Adjusted posting function to handle a promise asynchronously with async/await.   
Changed new message functions to functionally map to fresh state variables instead of mutating state.   
Messages sort with `Moment.js` functions at the fetch before mapping to message UI.  

## Further comments / Discussion Response

I was able to use the Cypress tests to track down the source of the bad POST requests. The issues with state were visible after logging input through each of the message functions, and I was able to track down the problems after confirming the issue in the console.
I used `Moment.js`'s `isBefore` function to sort timestamped messages, as `Moment` is already being used to generate timestamps.

### CODE REVIEW

File: client/src/components/ActiveChat/Messages.js [hunk 1]
```diff
@@ -8,7 +8,11 @@ const Messages = (props) => {
 
   return (
     <Box>
-      {messages.map((message) => {
+      {messages
```
Comment (xynree) [inline client/src/components/ActiveChat/Messages.js:11]: Sort and map functions here could be memoized for optimization. I simply chained onto what was already here for consistency.

File: client/src/components/ActiveChat/Messages.js [hunk 2]
```diff
@@ -8,7 +8,11 @@ const Messages = (props) => {
...
+        .sort((prev,curr) => 
+          moment(prev.createdAt).isBefore(curr.createdAt)? -1 : 1
+        )
+        .map((message) => {
```
Comment (uimarshall) [inline client/src/components/ActiveChat/Messages.js:15]: This only works on page reload. However, it's not very efficient, since this will get executed on every component re-render. Ideally, we want the messages to be sorted before they reach the state. See if you can sort the messages at the instance of fetching the conversations. Then, there won't be a need for memorization.

- Additionally, you don't have to add comments before the review so that it will be easy for the reviewers to read the codes.

File: client/src/components/Home.js [hunk 3]
```diff
@@ -54,17 +54,17 @@ const Home = ({ user, logout }) => {
...
 
-  const sendMessage = (data, body) => {
+  const sendMessage = useCallback((data, body) => {
     socket.emit("new-message", {
       message: data.message,
       recipientId: body.recipientId,
       sender: data.sender,
     });
-  };
+  },[socket]);
```
Comment (uimarshall) [inline client/src/components/Home.js:63]: These changes are not related to the issue being addressed. Please revert them.

File: client/src/components/Home.js [hunk 4]
```diff
@@ -54,17 +54,17 @@ const Home = ({ user, logout }) => {
...
 
-  const postMessage = (body) => {
+  const postMessage = async (body) => {
     try {
-      const data = saveMessage(body);
+      const data = await saveMessage(body);
```
Comment (uimarshall) [inline client/src/components/Home.js:67]: Great job fixing this bug! 💯

File: client/src/components/Home.js [hunk 5]
```diff
@@ -80,14 +80,14 @@ const Home = ({ user, logout }) => {
...
         if (convo.otherUser.id === recipientId) {
           convo.messages.push(message);
           convo.latestMessageText = message.text;
           convo.id = message.conversationId;
-        }
-      });
-      setConversations(conversations);
+        } 
+        return convo;
+      }));
```
Comment (uimarshall) [inline client/src/components/Home.js:90]: While this works to "fix" the bug by triggering a re-render, it doesn't really address the root cause of the issue. Take a look at the state update functions involved in the sending of a message, particularly at the way that the state update (dispatch) functions are handling updating the state (when the next state needs to be based on the previous state, a [funcitonal update](https://reactjs.org/docs/hooks-reference.html#functional-updates) is needed) 🙂.
Take a look at the `addOnlineUser` function to reference how to properly handle this.

File: client/src/components/Home.js [hunk 6]
```diff
@@ -105,13 +105,13 @@ const Home = ({ user, logout }) => {
...
+      setConversations(conversations.map((convo) => {
         if (convo.id === message.conversationId) {
           convo.messages.push(message);
           convo.latestMessageText = message.text;
-        }
-      });
-      setConversations(conversations);
+        } 
+        return convo;
+      }));
```
Comment (uimarshall) [inline client/src/components/Home.js:114]: Same as above 🙂

**Review by uimarshall — COMMENTED** [review]
Hi! @xynree,
Thank you for your work on this PR 👍🏽.
Good catch on all the issues 👍 .
Please address the comments below and let us know by using the `/hatchy pr` command on Slack and another reviewer will come by to give this PR another look and give you approval.

**A small request:** please don’t resolve the conversations, but feel free to reply to them.

**Review by Shruti-Dongre — APPROVED** [review]
Thank you for your work in this PR🙏.

Good job addressing all the above comments! Consider this PR approved and feel free to merge👍
